src/flatten_to_stream.rs

```diff
@@ -0,0 +1,101 @@
+use core::mem;
+use core::marker;
+use {Async, Future, Poll};
+use stream::Stream;
+
+/// Future for the `flatten_to_stream` combinator, flattening a future-of-a-stream to get just
+/// the result of the final future as a stream.
+///
+/// This is created by the `Future::flatten_to_stream` method.
+#[must_use = "streams do nothing unless polled"]
+pub struct FlattenToStream<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    state: FlattenToStreamState<F>
+}
+
+pub fn new<F>(f: F) -> FlattenToStream<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    FlattenToStream {
+        state: FlattenToStreamState::Future(f)
+    }
+}
+
+/// Stream that emits single error
+enum FailedState<T, E> {
```

Comment:
For the impl of `Stream` yeah, it's required, but it seems a little odd here? Also yes you can't have ownership differ on arms of `match`, but that's dealt with in a variety of ways throughout the futures repos.